src/typechecker.erl

```diff
@@ -4174,14 +4197,15 @@ add_type_pat(CONS = {cons, P, PH, PT}, ListTy, Env, VEnv) ->
             TailTy = normalize(type(union, [ListTy, type(nil)]), Env),
             {_TailPatTy, _TauUBound, VEnv3, Cs} = add_type_pat(PT, TailTy, Env, VEnv2),
             NonEmptyTy = rewrite_list_to_nonempty_list(ListTy),
-            {NonEmptyTy, NonEmptyTy, VEnv3, Cs};
+            {type(none), NonEmptyTy, VEnv3, Cs};
         {elem_ty, ElemTy, Cs1} ->
-            {_PatTy1, _UBound1, VEnv2, Cs2} =
+            {PatTy1, _UBound1, VEnv2, Cs2} =
                 add_type_pat(PH, normalize(ElemTy, Env), Env, VEnv),
             TailTy = normalize(type(union, [ListTy, type(nil)]), Env),
             {_PatTy2, _Ubound2, VEnv3, Cs3} = add_type_pat(PT, TailTy, Env, VEnv2),
+            PatTy = type(nonempty_list, [PatTy1]),
             NonEmptyTy = rewrite_list_to_nonempty_list(ListTy),
-            {NonEmptyTy, NonEmptyTy, VEnv3, constraints:combine([Cs1, Cs2, Cs3])};
+            {PatTy, NonEmptyTy, VEnv3, constraints:combine([Cs1, Cs2, Cs3])};
```

Comment:
This is hard do understand when you haven't looked at it for a while. :-)

I think some comments would be helpful, e.g.

* %% PatTy1 is the type exhausted by the cons-pattern head (PH)

* %% _PatTy2 is the type exhausted by the cons-pattern tail (PT), which is irrelevant (why??? example???)

* %% PatTy is the type exhausted by the cons pattern.

Reply:
To be honest, to figure it out I had to look at an execution trace of a specific minimal example. I wouldn't be able to even start without that. Only then names like `PH` started making sense. I was not sure either why `PatTy1` was ignored, given it turned out to be useful.

In this light:

> %% _PatTy2 is the type exhausted by the cons-pattern tail (PT), which is irrelevant (why??? example???)

I'm not 100% sure it's irrelevant, but it didn't seem so yet. Maybe we just haven't found an example where it's relevant. That's why I think self-gradualizing is an important step for the project - this entire PR based on false positives found thanks to `make gradualize`.

I'll add some comments.